android/modules/database/src/java/ti/modules/titanium/database/DatabaseModule.java

```diff
@@ -53,101 +53,108 @@ public DatabaseModule()
 			} else {
-				String name = TiConvert.toString(file);
-				SQLiteDatabase db = TiApplication.getInstance().openOrCreateDatabase(name, Context.MODE_PRIVATE, null);
+				throw new RuntimeException("SQLiteDatabase.openDatabase() returned null for path: " + absolutePath);
+			}
+		} else if (file instanceof String) {
+			String name = (String) file;
+			SQLiteDatabase db = TiApplication.getInstance().openOrCreateDatabase(name, Context.MODE_PRIVATE, null);
+			if (db != null) {
 				dbp = new TiDatabaseProxy(name, db);
+			} else {
+				throw new RuntimeException("SQLiteDatabase.openOrCreateDatabase() returned null for name: " + name);
 			}
-
-			Log.d(TAG, "Opened database: " + dbp.getName(), Log.DEBUG_MODE);
-
-		} catch (SQLException e) {
-			String msg = "Error opening database: " + dbp.getName() + " msg=" + e.getMessage();
-			Log.e(TAG, msg, e);
-			throw e;
+		} else if (file != null) {
+			throw new IllegalArgumentException("Ti.Database.open() argument must be of type 'String' or 'File'.");
+		} else {
+			throw new IllegalArgumentException("Ti.Database.open() was given a null argument.");
 		}
 
+		// Return a proxy to the opened database.
+		Log.d(TAG, "Opened database: " + dbp.getName(), Log.DEBUG_MODE);
 		return dbp;
 	}
 
 	@Kroll.method
 	public TiDatabaseProxy install(KrollInvocation invocation, String url, String name) throws IOException
 	{
-		try {
-			Context ctx = TiApplication.getInstance();
-			for (String dbname : ctx.databaseList()) {
-				if (dbname.equals(name)) {
-					return open(name);
-				}
+		// Validate arguments.
+		if ((url == null) || url.isEmpty()) {
+			throw new IllegalArgumentException("Ti.Database.install() 1st argument must be a non-empty string.");
+		}
+		if ((name == null) || name.isEmpty()) {
+			throw new IllegalArgumentException("Ti.Database.install() 2nd argument must be a non-empty string.");
+		}
+
+		// Do not continue if the database has already been installed.
+		// Open a connection to it and stop here.
+		Context ctx = TiApplication.getInstance();
+		for (String dbname : ctx.databaseList()) {
+			if (dbname.equals(name)) {
+				return open(name);
+			}
+		}
+
+		// Fetch a path to the source database file. This is the file to be copied/installed.
+		// Throw an exception if the source database was not found.
+		Log.d(TAG, "db url is = " + url, Log.DEBUG_MODE);
+		if (invocation == null) {
+			throw new RuntimeException("Ti.Database.install() was given a null 'KrollInvocation' object.");
+		}
+		TiUrl tiUrl = TiUrl.createProxyUrl(invocation.getSourceUrl());
+		String sourcePath = TiUrl.resolve(tiUrl.baseUrl, url, null);
+		TiBaseFile srcDb = TiFileFactory.createTitaniumFile(sourcePath, false);
+		if (srcDb.isFile() == false) {
+			String message = "Ti.Database.install() failed to find 1st argument's source database file: " + url;
+			throw new java.io.FileNotFoundException(message);
+		}
+
+		// open an empty one to get the full path and then close and delete it
```

Comment:
this comment seems to no longer describe reality here

Reply:
Admittedly, I haven't really changed this code block other than its indentation and added extra null checks.

Having a harder look at it now, this can be handled better. It should go through our `TiFileFactory` class which is where we handle all of our internal customer URL schemes if given. From this we can fetch the filesystem path if applicable.

Reply:
I've confirmed that `install()` method was missing support for URLs referencing files on the file system such as `file://` and  `appdata-private://`. The `open()` method didn't support any URLs file referencing schemes at all. I've updated both to support this and have already tested it on my end.